apps/test/utilityTests.js

```diff
@@ -449,3 +449,258 @@ describe('forceInsertTopBlock', function () {
+  });
+
+  it('replaces all &nbsp; with spaces', function () {
+    assert.equal(' ', unescapeText(' '));
+    assert.equal(' ', unescapeText('&nbsp;'));
+    assert.equal('  ', unescapeText(' &nbsp;'));
+    assert.equal('  ', unescapeText('&nbsp;&nbsp;'));
+    assert.equal('   ', unescapeText(' &nbsp; '));
+    assert.equal('    ', unescapeText(' &nbsp; &nbsp;'));
+  });
+
+  it('Unescapes pre-escaped content correctly', function () {
+    assert.equal('&amp; &nbsp; &gt; &lt;',
+        unescapeText('&amp;amp; &amp;nbsp; &amp;gt; &amp;lt;'));
+  });
+
+  it('Unwraps <div>-wrapped lines, adding line breaks', function () {
+    var input = 'Line 1<div>Line 2</div><div>Line 3</div>';
+    var expected = [
+      'Line 1',
+      'Line 2',
+      'Line 3'
+    ].join('\n');
+    assert.equal(expected, unescapeText(input));
+  });
+
+  it('If input starts with <div> treats that as the first line', function () {
+    var input = '<div>Line 1</div><div>Line 2</div>';
+    var expected = [
+      'Line 1',
+      'Line 2'
+    ].join('\n');
+    assert.equal(expected, unescapeText(input));
+  });
+
+  it('Converts <div><br></div> to a blank line', function () {
+    var input, expected;
+
+    input = '<div><br></div><div>Line 2</div><div>Line 3</div>';
+    expected = [
+      '',
+      'Line 2',
+      'Line 3'
+    ].join('\n');
+    assert.equal(expected, unescapeText(input), 'first line blank');
+
+    input = 'Line 1<div><br></div><div>Line 3</div>';
+    expected = [
+      'Line 1',
+      '',
+      'Line 3'
+    ].join('\n');
+    assert.equal(expected, unescapeText(input), 'second line blank');
+
+    input = 'Line 1<div>Line 2</div><div><br></div>';
+    expected = [
+      'Line 1',
+      'Line 2',
+      ''
+    ].join('\n');
+    assert.equal(expected, unescapeText(input), 'third line blank');
+
+    input = '<div><br></div><div><br></div><div><br></div>';
+    expected = [
+      '',
+      '',
+      ''
+    ].join('\n');
+    assert.equal(expected, unescapeText(input), 'all lines blank');
+  });
+
+  describe('is the inverse of escapeText', function () {
+    var escapeText = utils.escapeText;
+    var assertRoundTrip = function assertRoundTrip(testString) {
+      var intermediateForm = escapeText(testString);
+      assert.equal(testString, unescapeText(intermediateForm),
+          'Failed with intermediate form ' + intermediateForm);
+    };
+
+    it('With newline', function () {
+      assertRoundTrip('Line1\nLine2');
+    });
+
+    it('With leading newline', function () {
+      assertRoundTrip('\nLine1\nLine2');
+    });
+
+    it('With multiple leading newlines', function () {
+      assertRoundTrip('\n\n \nLine1\nLine2');
+    });
+
+    describe('fuzz tests (randomized)', function () {
+      var stringPool = [
+        '<div>',
+        '</div>',
+        '<br>',
+        '<a>',
+        '&',
+        '<',
+        '>',
+        ' ',
+        '&amp;',
+        '&lt;',
+        '&gt;',
+        '&nbsp;',
+        '\n'
```

Comment:
I'm open to suggestions here - I just threw in components that I know are likely to be problematic for this transformation.